src/components/VmActions/AutoSelectConsole.js

```diff
@@ -0,0 +1,133 @@
+    dispatchStatus: 'doNotDispatch',
+    active: false,
+  }
+
+  const [state, setState] = useState(initialState)
+  const {
+    dispatchStatus,
+    selectedProtocol,
+    active,
+    // failureReason,
+    // consoles,
+  } = state
+
+  useEffect(() => {
+    if (show) {
+      setState(state => ({
+        ...state,
+        dispatchStatus: 'shouldDispatch',
+        active: true,
+      }))
+      fetchConsoles(vm.get('id'))
+    }
+  }, [show]) // eslint-disable-line react-hooks/exhaustive-deps
+
+  useEffect(() => {
+    const newState = {}
+
+    const { selectedProtocol, failureReason } = selectProtocol(vm.get('consoles'), requestActive)
+
+    if (selectedProtocol || failureReason) {
+      newState.selectedProtocol = selectedProtocol
+      newState.failureReason = failureReason
+    }
+    if (active && selectedProtocol && dispatchStatus !== 'doNotDispatch') newState.dispatchStatus = 'dispatched' // cannot change from 'doNotDispatch' to 'dispatched'
+    if (active && !selectedProtocol && failureReason && !requestActive && dispatchStatus !== 'doNotDispatch') onFailure(failureReason)
+
+    setState(state => ({ ...state, ...newState }))
+  }, [vm, requestActive]) // eslint-disable-line react-hooks/exhaustive-deps
+
+  const onClose = () => {
+    if (dispatchStatus === 'dispatched') setState(state => ({ ...state, dispatchStatus: 'doNotDispatch', active: false })) // can only change from 'dispatched' to 'doNotDispatch'
+    props.onClose()
+  }
+
+  const selectProtocol = (consoles, requestActive) => {
+    let selectedProtocol, failureReason
+    if (!requestActive) {
+      if (consoles.size > 0) {
+        if (consoles.size === 1) {
+          selectedProtocol = consoles.first()
+        } else {
+          selectedProtocol = consoles.find(c => c.get('protocol') === defaultConsole)
+          failureReason = !selectedProtocol && 'noDefault'
+        }
+      } else {
+        failureReason = 'emptyConsolesList'
+      }
+    }
+    return { selectedProtocol, failureReason }
+  }
+
+  return (
+    <ConsoleConfirmationModal
+      consoleId={selectedProtocol ? selectedProtocol.get('id') : ''}
+      vm={vm}
+      show={active && !!selectedProtocol && dispatchStatus !== 'doNotDispatch'}
+      isNoVNC={!!selectedProtocol && (selectedProtocol.get('protocol') === 'vnc' ? isNoVNC : false)}
+      onClose={onClose}
+      id={id}
+    />
+  )
+}
+
+AutoSelectConsole.propTypes = {
+  vm: PropTypes.object.isRequired,
+  show: PropTypes.bool,
+  defaultConsole: PropTypes.string,
+  isNoVNC: PropTypes.bool,
+  fetchConsoles: PropTypes.func.isRequired,
+  requestActive: PropTypes.bool.isRequired,
+  onClose: PropTypes.func.isRequired,
+  onFailure: PropTypes.func.isRequired,
+  id: PropTypes.string.isRequired,
+}
+
+export default connect(
+  (state) => ({
+    defaultConsole: state.config.get('defaultConsole'),
+    requestActive: !state.activeRequests.isEmpty(),
+  }),
+  (dispatch, { defaultConsole }) => ({
```

Comment:
`defaultConsole` here is passed as `ownProps` but is retrieved from the store. Usually `ownProps` and store based are different categories.